### PR TITLE
Give `_runFlutterTest` the ability to validate command output

### PR DESCRIPTION
**Pull request by jamesderlin**
In another change (#37646), I want to test that a test fails and
prints expected output.  I didn't see an existing way to do that, so
I modified `_runFlutterTest` and `runCommand` to allow capturing the
output.  Currently capturing and printing output are mutually
exclusive since we don't need both.

Some awkward bits:
* There already exists a `runAndGetStdout` function that is very
  similar to `runCommand`, and this change makes the conceptual
  distinction more confusing.

* `runFlutterTest` has multiple code paths for different
  configurations.  I don't understand what the different paths are
  for, and I added output checking only along one of them.

## Tests

Tested manually.  Will be exercised by PR #37646.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: dev/bots/run_command.dart [hunk 1]
```diff
@@ -133,9 +142,12 @@ Future<void> runCommand(String executable, List<String> arguments, {
     if (failureMessage != null) {
       print(failureMessage);
     }
-    if (!printOutput) {
-      stdout.writeln(utf8.decode((await savedStdout).expand<int>((List<int> ints) => ints).toList()));
-      stderr.writeln(utf8.decode((await savedStderr).expand<int>((List<int> ints) => ints).toList()));
+
+    // Print the output when we get unexpected results (unless output was
+    // printed already).
+    if (outputMode != OutputMode.print) {
```
Comment (dnfield) [inline dev/bots/run_command.dart]: This should be a switch statement in case we end up adding more cases to the enum

File: dev/bots/run_command.dart [hunk 2]
```diff
@@ -110,7 +113,7 @@ Future<void> runCommand(String executable, List<String> arguments, {
     .where((String line) => removeLine == null || !removeLine(line))
     .map((String line) => '$line\n')
     .transform(const Utf8Encoder());
-  if (printOutput) {
+  if (outputMode == OutputMode.print) {
```
Comment (dnfield) [inline dev/bots/run_command.dart]: Should be a switch statement.

**Review by dnfield — COMMENTED** [review]
Couple nits

File: dev/bots/test.dart [hunk 3]
```diff
@@ -975,3 +1009,8 @@ Future<void> _androidGradleTests(String subShard) async {
     await _runDevicelabTest('module_host_with_custom_build_test', env: env);
   }
 }
+
+/// Returns true if `p` logically implies `q`, false otherwise.
+///
+/// If `p` is true, `q` must be true.
+bool _implies(bool p, bool q) => !p || q;
```
Comment (jonahwilliams) [inline dev/bots/test.dart]: I don't think this function is carrying its weight here.

File: dev/bots/run_command.dart [hunk 4]
```diff
@@ -133,9 +142,12 @@ Future<void> runCommand(String executable, List<String> arguments, {
...
       print(failureMessage);
     }
-    if (!printOutput) {
-      stdout.writeln(utf8.decode((await savedStdout).expand<int>((List<int> ints) => ints).toList()));
-      stderr.writeln(utf8.decode((await savedStderr).expand<int>((List<int> ints) => ints).toList()));
+
+    // Print the output when we get unexpected results (unless output was
+    // printed already).
+    if (outputMode != OutputMode.print) {
+      stdout.writeln(flattenToString(await savedStdout));
```
Comment (jonahwilliams) [inline dev/bots/run_command.dart]: This change seems unrelated

Comment (jamesderlin) [reply dev/bots/run_command.dart]: Which part of it?  The `flattenToString` business is there because I wanted to reuse the same transformation that this code path was using (lines 133-134), and I didn't want to duplicate it.

File: dev/bots/test.dart (hunk 3 above)
Comment (jamesderlin) [reply dev/bots/test.dart]: I think it makes the assertion more readable and makes its intent clearer, but I can remove it if you like.

Comment (jonahwilliams) [reply dev/bots/test.dart]: Please remove

**Comment by jamesderlin** [comment]
@dnfield @jonahwilliams Could you take another look?

File: dev/bots/test.dart [hunk 5]
```diff
@@ -802,13 +808,17 @@ Future<void> _runFlutterTest(String workingDirectory, {
...
   bool printOutput = true,
+  OutputChecker outputChecker,
   List<String> options = const <String>[],
   bool skip = false,
   Duration timeout = _kLongTimeout,
   bq.TabledataResourceApi tableData,
   Map<String, String> environment,
   List<String> tests = const <String>[],
 }) async {
+  // Support printing output or capturing it for matching, but not both.
```
Comment (dnfield) [inline dev/bots/test.dart]: Something like this should be the message on the assert rather than a comment, e.g. `assert(!printOutput || outputChecker == null, 'Output can either be captured or matched, but not both')`

File: dev/bots/run_command.dart [hunk 6]
```diff
@@ -83,12 +83,15 @@ Future<void> runCommand(String executable, List<String> arguments, {
...
   String failureMessage,
-  bool printOutput = true,
+  OutputMode outputMode = OutputMode.print,
+  CapturedOutput output,
   bool skip = false,
   bool expectFlaky = false,
   Duration timeout = _kLongTimeout,
   bool Function(String) removeLine,
 }) async {
+  assert((outputMode == OutputMode.capture) == (output != null));
```
Comment (dnfield) [inline dev/bots/run_command.dart]: Please add a message for this - e.g. `assert((outputMode == OutputMode.capture) == (output != null), 'The output parameter must not be null if outputMode == OutputMode.capture, and must be null otherwise')`.

Alternatively, perhaps this should be some kind of class itself rather than an enum and a callback...

File: dev/bots/test.dart (hunk 5 above)
Comment (jamesderlin) [reply dev/bots/test.dart]: Oops.  Fixed.

File: dev/bots/run_command.dart (hunk 6 above)
Comment (jamesderlin) [reply dev/bots/run_command.dart]: Fixed.  I think that keeping the enum and assert is a bit simpler for now.

**Review by dnfield — APPROVED** [review]
LGTM

File: dev/bots/run_command.dart [hunk 7]
```diff
@@ -147,3 +170,18 @@ Future<void> runCommand(String executable, List<String> arguments, {
     exit(1);
   }
 }
+
+T identity<T>(T x) => x;
```
Comment (jonahwilliams) [inline dev/bots/run_command.dart]: This is only used in a single place, just use a closure

File: dev/bots/run_command.dart [hunk 8]
```diff
@@ -147,3 +170,18 @@ Future<void> runCommand(String executable, List<String> arguments, {
...
+
+/// Flattens a nested list of UTF-8 code units into a single string.
+String flattenToString(List<List<int>> chunks) =>
+  utf8.decode(chunks.expand<int>(identity).toList(growable: false));
```
Comment (jonahwilliams) [inline dev/bots/run_command.dart]: What does growable: false do here? If not required I would remove it.

File: dev/bots/run_command.dart [hunk 9]
```diff
@@ -147,3 +170,18 @@ Future<void> runCommand(String executable, List<String> arguments, {
     exit(1);
   }
 }
+
+T identity<T>(T x) => x;
+
+/// Flattens a nested list of UTF-8 code units into a single string.
+String flattenToString(List<List<int>> chunks) =>
```
Comment (jonahwilliams) [inline dev/bots/run_command.dart]: This is only used in this file so make it private

**Review by jonahwilliams — APPROVED** [review]
LGTM with nits

File: dev/bots/run_command.dart (hunk 7 above)
Comment (jamesderlin) [reply dev/bots/run_command.dart]: Okay, I'll remove it, but I think having it is more readable than:

`utf8.decode(chunks.expand<int>((List<int> ints) => ints).toList());`

with all of the nested parentheses.

File: dev/bots/run_command.dart (hunk 9 above)
Comment (jamesderlin) [reply dev/bots/run_command.dart]: Fixed.
